util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+        Log(NULL, "leveldb HotBackup unable to delete trigger file %s", GetTriggerPath());
+        Log(NULL, "leveldb HotBackup now disabled until program restart");
+    }   // if
+
+    return;
+
+}   // ResetHotBackupTrigger
+
+
+/**
+ * A database instance marks its intent to backup
+ */
+void
+HotBackup::HotBackupScheduled()
+{
+    inc_and_fetch(&m_JobsPending);
+
+    return;
+
+}   // HotBackupScheduled
+
+
+/**
+ * A database instance marks that its backup completed
+ */
+void
+HotBackup::HotBackupFinished()
+{
+    int ret_val;
+
+    ret_val=dec_and_fetch(&m_JobsPending);
+
+    // 1 means CheckHotBackupTrigger()'s counter is only
+    //   one left
+    if (1==ret_val)
+    {
+        // hmm, call file sync() at this point?
+        Log(NULL, "leveldb HotBackup complete.");
+        ResetTrigger();
+    }   // if
+
+    return;
+
+}   // HotBackupFinished
+
+
+/**
+ * Routine called by DBList()->ScanDBs.  This has full context of one open
+ *  leveldb database.  Validates then schedules a backup job via compaction
+ *  threads.
+ */
+void
+DBImpl::HotBackup()
+{
+    bool create_backup_event(false);
+
+    // hold mutex while reviewing database state
+    //  (note:  never good to call Log() while holding mutex
+    //   but these are unlikely error cases)
+    {
+        // get mutex,
+        MutexLock l(&mutex_);
+
+        // set backup flag && see if
+        assert(false==hotbackup_pending_);
+        if (!hotbackup_pending_)
+        {
+            // test if shutting_down,
+            if (!shutting_down_.Acquire_Load())
+            {
+                create_backup_event=true;
+                hotbackup_pending_=true;
+            }   // if
+            else
+            {
+                Log(options_.info_log, "HotBackup aborted, shutdown in progress");
+            }   // else
+        }   // if
+        else
+        {
+            Log(options_.info_log, "HotBackup aborted, backup already in progress");
+        }   // else
+    }   // mutex released
+
+    // post event and log message outside mutex
+    if (create_backup_event)
+    {
+        // increment pending backup count
+        gHotBackup->HotBackupScheduled();
+
+        // schedule backup job within compaction queue
+        ThreadTask * task=new HotBackupTask(this);
+        gCompactionThreads->Submit(task, true);
+        Log(options_.info_log, "HotBackup queued");
+    }   // if
+
+    return;
+
+}   // DBImpl::HotBackup
+
+
+void
+HotBackupTask::operator()()
+{
+    bool good;
+    long log_position(0);
+
+    // rotate directories (tiered storage reminder)
+    good=gHotBackup->PrepareDirectories(m_DBImpl.GetOptions());
+
+    // purge imm or current write buffer
+    good=good && m_DBImpl.PurgeWriteBuffer();
+
+    // Gather random data items for MANIFEST and
+    //  other supporting files.
+    if (good)
+    {
+        // get LOG file size
+        if (NULL!=m_DBImpl.GetLogger())
+            log_position=m_DBImpl.GetLogger()->LogSize();
```

Comment:
For the default Logger (PosixLogger), the LogSize() method calls ftell(), which returns -1 on error, so log_position should be checked before using it below. Either that, or change the conditional below from "0!=log_position" to "0<log_position", if doing everything else still makes sense.